### PR TITLE
chore: add full API documentation

### LINKED ISSUE #3: 🚀 [DOC]: add full API documentation
### 👀 Background

We need to expose a full documentation of the current API.

### ✌️ What is the actual behavior?

No API doc !

### 🕵️‍♀️ How to reproduce the current behavior?

_No response_

### 🤞 What is the expected behavior?

Have an API doc 🏆 !

### 😎 Proposed solution.

_No response_

### 🙏 Would you be willing to submit a PR?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tony-go**
## Change description

Add full API documentation.

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

This PR resolves #3. 

### Development

- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [x] Automated tests covering modified code pass

### Security

- [x] Security impact of change has been considered
- [x] Code follows project security practices and guidelines

### Code review 

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to issue tracker where applicable

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# API
+
+If you're familiar with TypeScript you could check the `index.d.ts`.
+
+Where `const pf = new PluginFactory("your-plugin-name")`;
+
+## pf.name(): string
```
Comment (olivierodo) [inline API.md:7]: From what i see in the code name is not a method but  a getter accessible through `pf.name`

File: API.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+Return the plugin name.
+
+## pf.addState(key: string, value: any): PluginFactory
+
+Add a state property to the current state. You can consume it in your handlers using this API:
+```js
+cons state = this[pf.name];
```
Comment (olivierodo) [inline API.md:15]: This has been revert back by the original implementation that you proposed `this.state.key`

File: API.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+## pf.getConfig(): Config
+
```
Comment (olivierodo) [inline API.md:19]: Additionally, if the user try to access to the config from the handler he could use the method: `this.getConfig('pluginName')` 

ref: https://github.com/restqa/restqa/blob/89cfd88f93721e52067fc563135a107ab4ff234e/src/core/bootstrap.js#L132

File: API.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+Add a Then step.
+
+## pf.addBeforeHook(handler: (...args: any[]): PluginFactory
+
+Add a Before hook.
+
+If you want to use tags you could use these interfaces:
+- pf.addBeforeHook(tags: string, handler: (...args: any[]): PluginFactory
+- pf.addBeforeHook(options: { tags: string }, handler: (...args: any[]): PluginFactory
```
Comment (olivierodo) [inline API.md:40]: Maybe add the link to cucumberJS doc ?

**Review by olivierodo — CHANGES_REQUESTED** [review]
Small changes required.
Another important thing to mention in to the doc is: All the handler function can't be  arrow function otherwise the handler cant access to the `this` apis.

File: API.md (hunk 1 above)
Comment (tony-go) [reply API.md:7]: Right ^^

File: API.md (hunk 3 above)
Comment (tony-go) [reply API.md:19]: User can use `pf.getConfig()` from the handler too. I think we should favor one API for one usage.

File: API.md (hunk 4 above)
Comment (tony-go) [reply API.md:40]: Already in the README.md, but yeah could be nice to put it there too.

**Pull request merged**